mmdet/utils/memory.py

```diff
@@ -0,0 +1,217 @@
+        3. Since the function might be called more than once, it has to be
+            stateless.
+    """
+
+    def __init__(self,
+                 keep_type=True,
+                 convert_cpu=True,
+                 return_gpu=True,
+                 test=False,
+                 return_fp16=True):
+        self.logger = get_root_logger()
+        self.keep_type = keep_type
+        self.convert_cpu = convert_cpu
+        self.return_gpu = return_gpu
+        self.test = test
+        self.return_fp16 = return_fp16
+
+    def retry_if_cuda_oom(self, func):
+        """Makes a function retry itself after encountering pytorch's CUDA OOM
+        error. It will do the following steps:
+
+            1. first retry after calling `torch.cuda.empty_cache()`.
+            2. If that still fails, it will then retry by trying to
+              convert inputs to FP16.
+            3. If that still fails trying to convert inputs to CPUs.
+              In this case, it expects the function to dispatch to
+              CPU implementation.
+
+        Args:
+            func: a stateless callable that takes tensor-like objects
+                as arguments.
+        Returns:
+            func: a callable which retries `func` if OOM is encountered.
+        """
+
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+
+            # raw function
+            if not self.test:
+                with _ignore_torch_cuda_oom():
+                    return func(*args, **kwargs)
+
+                # Clear cache and retry
+                torch.cuda.empty_cache()
+                with _ignore_torch_cuda_oom():
+                    return func(*args, **kwargs)
+
+            # get the type and device of first tensor
+            dtype, device = None, None
+            values = args + tuple(kwargs.values())
+            for value in values:
+                if isinstance(value, torch.Tensor):
+                    dtype = value.dtype
+                    device = value.device
+                    break
+            if dtype is None or device is None:
+                raise ValueError('There is no tensor in the inputs, '
+                                 'cannot get dtype and device.')
+
+            # Try to use FP16
+            fp16_args = convert_tensor_type(args, dst_type=torch.half)
+            fp16_kwargs = convert_tensor_type(kwargs, dst_type=torch.half)
+            self.logger.info(f'Attempting to copy inputs of {str(func)} '
+                             f'to fp16 due to CUDA OOM')
+
+            if self.keep_type:
+                # get input tensor type, the output type will same as
+                # the first parameter type.
+                with _ignore_torch_cuda_oom():
+                    self.logger.info(f'Trying to convert output to {dtype}')
+                    output = func(*fp16_args, **fp16_kwargs)
+                    output = convert_tensor_type(
+                        output, src_type=torch.half, dst_type=dtype)
```

Comment:
we should have function in mmcv to convert tensors to fp16.

Reply:
yes,  the function is `mmcv.runner.cast_tensor_type`. But we also need to convert tensors from GPU to CPU or from CPU to GPU. MMCV does not have a related function. So we write a new function that can both covert device and dtype. 
I did not use `mmcv.runner.cast_tensor_type` because the new function is more intuitive for users